log4j-api/src/main/java/org/apache/logging/log4j/spi/Provider.java

```diff
@@ -97,127 +217,266 @@ public String getVersions() {
+        final String threadContextMapClass = props.getStringProperty(THREAD_CONTEXT_MAP_PROPERTY);
+        if (threadContextMapClass != null) {
+            return threadContextMapClass;
+        }
+        // Default based on properties
+        if (Constants.ENABLE_THREADLOCALS) {
+            return props.getBooleanProperty(GC_FREE_THREAD_CONTEXT_PROPERTY)
+                    ? GC_FREE_THREAD_CONTEXT_PROPERTY
+                    : COPY_ON_WRITE_CONTEXT_MAP;
+        }
+        return WEB_APP_CONTEXT_MAP;
     }
 
     /**
-     * Loads the {@link org.apache.logging.log4j.spi.ThreadContextMap} class specified by this Provider.
+     * Loads the {@link ThreadContextMap} class specified by this Provider.
      *
-     * @return the ThreadContextMap implementation class or {@code null} if there was an error loading it
+     * @return the {@code ThreadContextMap} implementation class or {@code null} if unspecified or a loading error
+     * occurred.
+     * @see #createThreadContextMap()
      */
     public Class<? extends ThreadContextMap> loadThreadContextMap() {
         if (threadContextMapClass != null) {
             return threadContextMapClass;
         }
-        if (threadContextMap == null) {
-            return null;
-        }
+        final String threadContextMap = getThreadContextMap();
         final ClassLoader loader = classLoader.get();
-        if (loader == null) {
+        // Support for deprecated {@code META-INF/log4j-provider.properties} format.
+        // In the remaining cases {@code loader == null}.
+        if (loader == null || threadContextMap == null) {
             return null;
         }
         try {
             final Class<?> clazz = loader.loadClass(threadContextMap);
             if (ThreadContextMap.class.isAssignableFrom(clazz)) {
                 return clazz.asSubclass(ThreadContextMap.class);
+            } else {
+                LOGGER.error(
+                        "Class {} specified in {} does not extend {}",
+                        threadContextMap,
+                        getUrl(),
+                        ThreadContextMap.class.getName());
             }
         } catch (final Exception e) {
-            LOGGER.error("Unable to create class {} specified in {}", threadContextMap, url.toString(), e);
+            LOGGER.error("Unable to load class {} specified in {}", threadContextMap, url.toString(), e);
         }
         return null;
     }
 
+    /**
+     * Extension point for providers to create a {@link ThreadContextMap}
+     * <p>
+     *     The default implementation:
+     * </p>
+     * <ol>
+     *     <li>calls {@link #loadThreadContextMap},</li>
+     *     <li>if the previous call returns {@code null}, it calls {@link #getThreadContextMap} to instantiate one of
+     *     the internal implementations,</li>
+     *     <li>it returns a no-op map otherwise.</li>
+     * </ol>
+     * @since 2.24.0
+     */
+    protected ThreadContextMap createThreadContextMap() {
```

Comment:
Similar to `createLoggerContextFactory()`, I don't understand why this needs to have `public` visibility.